File: src/com/kingmang/lazurite/core/Classes.java
package com.kingmang.lazurite.core;

import com.kingmang.lazurite.exceptions.LZRException;
import com.kingmang.lazurite.runtime.ClassInstanceValue;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.HashMap;
import java.util.Map;

@NoArgsConstructor
public final class Classes {

    @Getter
    private static final Map<String, ClassInstanceValue> classes;
    static {
        classes = new HashMap<>();
    }

    public static void clear() {
        classes.clear();
    }

    public static boolean isExists(String key) {
        return classes.containsKey(key);
    }
    
    public static ClassInstanceValue get(String key) {
        if (!isExists(key)) throw new LZRException("UnknownClassException ", "Unknown class " + key);
        return classes.get(key);
    }
    
    public static void set(String key, ClassInstanceValue classDef) {
        classes.put(key, classDef);
    }
}

File: src/main/java/com/kingmang/lazurite/core/Classes.kt
package com.kingmang.lazurite.core

import com.kingmang.lazurite.exceptions.LzrException
import com.kingmang.lazurite.runtime.ClassInstanceValue
import lombok.Getter
import lombok.NoArgsConstructor

object Classes {
    private val classes: MutableMap<String, ClassInstanceValue> = HashMap()

    @Synchronized
    fun clear() =
        classes.clear()

    @Synchronized
    fun isExists(key: String): Boolean =
        classes.containsKey(key)

    @Synchronized
    operator fun set(key: String, classDef: ClassInstanceValue) {
        classes[key] = classDef
    }

    @Synchronized
    operator fun get(key: String): ClassInstanceValue? {
        if (!isExists(key))
            throw LzrException("UnknownClassException ", "Unknown class $key")
        return classes[key]
    }
}
